File: Backend/Authenticate/src/main/java/com/shipfast/auth/entity/RoleRequest.java
package com.shipfast.auth.entity;

import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Lob;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Data;

@Data
@Entity
@Table(name = "role_request")
public class RoleRequest {

    @Id
    @Column(name = "request_id", nullable = false, updatable = false)
    private String requestId;

    @Column(name = "user_id", nullable = false)
    private String userId;

    @Column(nullable = false)
    private String email;

    @Column(nullable = false)
    private String name;

    @Column(name = "current_role", nullable = false)
    private String currentRole;

    @Column(name = "requested_role", nullable = false)
    private String requestedRole;

    @Lob
    private String reason;

    @Lob
    @Column(name = "agent_details_json")
    private String agentDetailsJson;

    @Lob
    @Column(name = "documents_json")
    private String documentsJson;

    @Column(nullable = false)
    private String status;

    @Column(name = "reviewed_by")
    private String reviewedBy;

    @Column(name = "reviewed_at")
    private LocalDateTime reviewedAt;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: mern-backend/src/models/RoleRequest.js
import { createModel } from '../db/mongo.js';

const RoleRequestMethods = {
  toDto: function () {
    return {
      id: this.requestId,
      requestId: this.requestId,
      userId: this.userId,
      userEmail: this.userEmail || '',
      userName: this.userName || '',
      requestedRole: this.requestedRole,
      reason: this.reason || '',
      status: this.status || 'PENDING',
      reviewedBy: this.reviewedBy || null,
      reviewedAt: this.reviewedAt || null,
      comments: this.comments || '',
      phone: this.phone || '',
      vehicleType: this.vehicleType || '',
      vehicleNumber: this.vehicleNumber || '',
      licenseNumber: this.licenseNumber || '',
      experience: this.experience || '',
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
};

const RoleRequestDefaults = {
  userEmail: '',
  userName: '',
  reason: '',
  status: 'PENDING',
  reviewedBy: null,
  reviewedAt: null,
  comments: '',
  phone: '',
  vehicleType: '',
  vehicleNumber: '',
  licenseNumber: '',
  experience: ''
};

export const RoleRequest = createModel('rolerequests', RoleRequestMethods, RoleRequestDefaults);
